src/google_sheets.rs

```diff
@@ -0,0 +1,247 @@
+use serde::{Deserialize, Serialize};
+
+/// Endpoint for the Google Sheets API.
+const BASE_ENDPOINT: &str = "https://sheets.googleapis.com/v4/";
+
+pub trait IntoSheetEntry {
+    fn into_sheet_entry(&self) -> Vec<String>;
+}
+
+#[derive(Serialize)]
+pub struct EmptyBody {}
+
+pub struct Sheets {
+    token: AccessToken,
+    client: Client,
+    sheet_id: String,
+}
+
+impl Sheets {
+    pub fn get_link_to_sheet(sheet_id: &str) -> String {
+        format!("https://docs.google.com/spreadsheets/d/{}/", sheet_id)
+    }
+    pub async fn initialize(sheet_id: &str) -> Result<Self, APIError> {
+        let token = match Sheets::authenticate().await {
+            Ok(t) => t,
+            Err(e) => return Err(e),
+        };
+
+        Sheets::new(token, sheet_id)
+    }
+
+    pub async fn authenticate() -> Result<AccessToken, APIError> {
+        // Read application secret from a file. Sometimes it's easier to compile it directly into
+        // the binary. The `client_secret` file contains JSON like `{"installed":{"client_id": ... }}`
+        let secret = oauth::read_application_secret("client_secret.json")
+            .await
+            .expect("client_secret.json");
+
+        // All authentication tokens are persisted to a file named `tokencache.json`.
+        // The authenticator takes care of caching tokens to disk and refreshing tokens once they've expired.
+        let auth = match InstalledFlowAuthenticator::builder(
+            secret,
+            InstalledFlowReturnMethod::HTTPRedirect,
+        )
+        .persist_tokens_to_disk("tokencache.json")
+        .build()
+        .await
+        {
+            Ok(a) => a,
+            Err(e) => return Err(APIError::from(e)),
+        };
+
+        let scope = &["https://www.googleapis.com/auth/spreadsheets"];
+        match auth.token(scope).await {
+            Ok(tok) => Ok(tok),
+            Err(err) => panic!("Could not authenticate properly {:?}", err),
+        }
+    }
+
+    pub fn new(token: AccessToken, sheet_id: &str) -> Result<Self, APIError> {
+        match Client::builder().build() {
+            Ok(client) => Ok(Self {
+                token,
+                client,
+                sheet_id: String::from(sheet_id),
+            }),
+            Err(_) => Err(APIError {
+                status_code: StatusCode::from_u16(500).unwrap(),
+                body: "Could not instantiate client".to_string(),
+            }),
+        }
+    }
+
+    pub fn request<T>(
+        &self,
+        method: Method,
+        path: &str,
+        body: T,
+        query_params: Option<Vec<(&str, &str)>>,
+    ) -> Request
+    where
+        T: Serialize,
+    {
+        // confirm URL can parse before continuing
+        let url = Url::parse(BASE_ENDPOINT).unwrap().join(&path).unwrap();
+
+        // TODO: use `self.token = Sheets::authenticate().await;` to attempt to update token from cache
+        // NOTE: this would make `request` async and I don't know if we want that or not
```

Comment:
Went to go change this and realized the reason this was a challenge was less due to the async nature and more due to the assignment to `self.token`.

This would require a mutable reference to `&mut self` for nearly every method in `google_sheets::Sheets`, which would require client code to use `let mut sheets = Sheets::initialize(&sheet_id).await.unwrap()` any time they instantiated it... which maybe that's fine, I'm unsure... but it felt "bad" to me for some reason.

Reply:
we also may not even need this given the way that every sheets instance is set up...